monitors/src/main/java/coppercore/monitors/Monitor.java

```diff
@@ -0,0 +1,205 @@
+
+public class Monitor {
+    String name; // Name to log the status of the monitor under, used by MonitoredSubsystem
+    boolean sticky; // Should the monitor still report a fault after conditions return to normal?
+    double timeToFault; // How long the value can be unacceptable before a fault occurs
+    BooleanSupplier isStateValid; // Supplier with which to check whether the value is acceptable
+    Runnable faultCallback; // Function to call when the fault happens
+
+    double triggeredTime = 0.0; // Timestamp when monitor was first triggered
+    // If this value is zero, the monitor has been triggered for less than 1 tick
+
+    boolean triggered = false; // Is the value currently unnacceptable?
+    boolean faulted = false; // Has the monitor detected a fault?
+
+    /**
+     * Creates a fault Monitor. This constructor takes all parameters at once. There is also a
+     * builder pattern supplied under MonitorBuilder. Using the builder is recommended because it
+     * makes code much more readable, but is not required.
+     *
+     * @param name the name of the monitor, which will be used by MonitoredSubsystem for logging.
+     * @param sticky whether the fault should remain faulted after conditions return to an
+     *     acceptable state.
+     * @param isStateValid supplier for whether the state is CURRENTLY valid. This doesn't need to
+     *     handle persistence, the monitor class will handle this automatically.
+     * @param timeToFault the time, in seconds, that isStateValid must return false before the a
+     *     fault is triggered.
+     * @param faultCallback a function called on every periodic loop while the monitor is in a
+     *     faulted state.
+     * @see MonitorBuilder
+     */
+    public Monitor(
+            String name,
+            boolean sticky,
+            BooleanSupplier isStateValid,
+            double timeToFault,
+            Runnable faultCallback) {
+        this.name = name;
+        this.sticky = sticky;
+        this.timeToFault = timeToFault;
+        this.isStateValid = isStateValid;
+
+        this.faultCallback = faultCallback;
+    }
+
+    /**
+     * This is the "main loop" of the Monitor. This should be called each in each periodic loop.
+     *
+     * @param currentTimeSeconds the current timestamp in seconds. This doesn't need to have a
+     *     specific frame of reference, it is used to detect when conditions have been unnacceptable
+     *     for enough time to fault.
+     */
+    public void periodic(double currentTimeSeconds) {
+        // currentTimeSeconds doesn't need to be from a specific point in history
+        // As long as the reference point is always the same, it could be from
+        // the robot being turned on, initialized, etc.
+
+        triggered = !isStateValid.getAsBoolean();
+        if (triggered) {
+            if (triggeredTime == 0.0) {
+                triggeredTime = currentTimeSeconds;
+            }
```

Comment:
Coding practice: don't check floating precision numbers for equality

This is sort of a nit because it is comparing to 0.0. I would prefer if triggeredTime was a negative number by default and we compared with <= though.

Reply:
What sentinel value would you prefer that I pick to indicate that it isn't currently triggered in place of zero?

Reply:
-1 is good